ocp_resources/machineset.py

```diff
@@ -0,0 +1,133 @@
+from ocp_resources.resource import NamespacedResource
+
+
+class MachineSet(NamespacedResource):
+    """
+    Machineset object.
+
+    Args:
+        cluster_name (str): OpenShift cluster name.
+        machine_role (str): machine role. e.g.: 'worker'.
+        machine_type (str): machine role. e.g.: 'worker'.
+        replicas (int): amount of replicas the machineset will have.
+        provider_spec (dict): provider spec information.
+
+        provider spec example:
+
+    {
+        "value": {
+            "apiVersion": "ovirtproviderconfig.machine.openshift.io/v1beta1",
+            "auto_pinning_policy": "none",
+            "cluster_id": 5612af70-f4f5-455d-b7df-fbad66accc38,
+            "cpu": {
+                "cores": 8,
+                "sockets": 1,
+                "threads": 1
+            },
+            "credentialsSecret": {
+                "name": "ovirt-credentials"
+            },
+            "kind": "OvirtMachineProviderSpec",
+            "memory_mb": 16000,
+            "os_disk": {
+                "size_gb": 31
+            },
+            "template_name": "ge2n1-gcwmg-rhcos",
+            "type": "server",
+            "userDataSecret": {
+                "name": "worker-user-data"
+            }
+        }
+    }
+    """
+
+    api_group = NamespacedResource.ApiGroup.MACHINE_OPENSHIFT_IO
+
+    def __init__(
+        self,
+        cluster_name=None,
+        name=None,
+        namespace=None,
+        teardown=True,
+        client=None,
+        machine_role="worker",
+        machine_type="worker",
+        replicas=1,
+        provider_spec=None,
+        yaml_file=None,
+    ):
+        super().__init__(
+            name=name,
+            namespace=namespace,
+            client=client,
+            teardown=teardown,
+            yaml_file=yaml_file,
+        )
+        self.replicas = replicas
+        self.cluster_name = cluster_name
+        self.machine_role = machine_role
+        self.machine_type = machine_type
+        self.provider_spec = provider_spec or {}
+
+    @property
+    def available_replicas(self):
+        return self.instance.status.availableReplicas
+
+    @property
+    def ready_replicas(self):
+        return self.instance.status.readyReplicas
+
+    @property
+    def desired_replicas(self):
+        return self.instance.status.replicas
+
+    @property
+    def provider_spec_value(self):
+        return self.instance.spec.template.spec.providerSpec.value
+
+    def to_dict(self):
+        res = super().to_dict()
+        if self.yaml_file:
+            return res
+
+        _spec, _selector, _metadata, _labels, _template, _provider_spec = (
+            "spec",
+            "selector",
+            "metadata",
+            "labels",
+            "template",
+            "providerSpec",
+        )
+
+        res[_metadata].setdefault(
```

Comment:
```
res[_metadata].setdefault(_labels, {}).update(
    {
                f"{self.api_group}/cluster-api-cluster": self.cluster_name,
                f"{self.api_group}/cluster-api-machine-role": self.machine_role,
                f"{self.api_group}/cluster-api-machine-type": self.machine_type,
            },
})
```

Reply:
what's the difference those two statements?

res[_metadata].setdefault(
            _labels,
            {
                f"{self.api_group}/cluster-api-cluster": self.cluster_name,
                f"{self.api_group}/cluster-api-machine-role": self.machine_role,
                f"{self.api_group}/cluster-api-machine-type": self.machine_type,
            },
        )

This statement will only overrun the labels in case the "lables" key does not appear in the dict.
So if the user applied its own labels then i will not overrun it.

Reply:
you want to update the existing one if any.